mds: add `drop cache` command

rishabh-d-dave opened the pull request:
Adds the command `drop cache` for MDS and an extra message so that client acknowledge to MDS that cache has been released. See - http://tracker.ceph.com/issues/23362

tchaikov commented:
retest this please.

tchaikov commented:
retest this please

tchaikov commented:
@rishabh-d-dave this change does not compile

rishabh-d-dave commented:
> @rishabh-d-dave this change does not compile

@tchaikov Made a mistake while cleaning up the debug code. Fixed it. Thanks!

batrick commented:
Please annotate the commit which fixes/resolves a Ceph tracker issue with:

`Fixes: http://tracker.ceph.com/issues/...`

This is essential when examining the history of the repository (this commit fixes what) and helps merge scripts identify issues that have been resolved by a merge.

rishabh-d-dave commented on src/mds/MDSRank.cc:
```diff
@@ -2983,11 +2983,71 @@ bool MDSRankDispatcher::handle_command(
+  // Optionally, flush the journal -
+  Formatter *fmtr = new JSONFormatter(true);
+  command_flush_journal(fmtr);
+  // TODO: do we want output from `flush journal`?
```
Running `ceph tell mds.a drop cache` displays output from `MDSRank::command_flush_journal()` as well as from `MDSRank::cache_status()`.
```
{
    "message": "",
    "return_code": 0
}
{
    "pool": {
        "items": 482,
        "bytes": 52398
    }
}
```
Do we want both or only the latter?

batrick replied:
both, combine into one object

batrick replied:
should be:

````
{
    "message": "",
    "return_code": 0
    "cache_status": {
      "pool": {
          "items": 482,
          "bytes": 52398
      }
  }
}
````

rishabh-d-dave replied:
Okay. Personally, I find it difficult to understand the meaning of first two lines. Will it be better if I make them a part of "flush journal" section like the following snippet?

```
{
    "flush_journal": {
        "message": "",
        "return_code": 0
    },
    "cache": {
        "pool": {
            "items": 482,
            "bytes": 52398
        }
    }
}
```

rishabh-d-dave replied:
Updating the PR with "flush_journal" part included. Let me know if you think it's not good idea.

rishabh-d-dave commented on src/mds/MDSRank.cc:
```diff
@@ -2240,7 +2240,7 @@ int MDSRank::_command_flush_journal(std::stringstream *ss)
   assert(ss != NULL);
 
-  Mutex::Locker l(mds_lock);
+  mds_lock.Lock();
```
@batrick I don't know if this change is really appropriate but without it the MDS crashes after the executing the command at this line [1] since it expects MDS to be locked.

[1] https://github.com/ceph/ceph/blob/master/src/mds/MDSDaemon.cc#L1143

batrick replied:
I'd suggest moving `Mutex::Locker l(mds_lock)` to surround the call to `_command_flush_journal` in `command_flush_journal`:

https://github.com/ceph/ceph/blob/189192b2f60429e95ff798314e71c4252cc6bf7f/src/mds/MDSRank.cc#L2225

to be

````
{
  Mutex::Locker l(mds_lock);
  const int r = _command_flush_journal(&ss);
}
````

Then comment `MDSrank::_command_flush_journal` that it expects `mds_lock` to be locked.

batrick commented on src/common/options.cc:
```diff
@@ -6938,6 +6938,12 @@ std::vector<Option> get_mds_options() {
+     .set_default(30)
+     .set_description("how long should the MDS wait for receiving "
+		      "CEPH_SESSION_RECALL_STATEMSG_ACK from all sessions "
+		      "while dropping cache."),
```
I think we can keep this vague. Don't mention dropping cache. Also, rename to `mds_session_recall_ack_timeout`.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -2983,11 +2983,71 @@ bool MDSRankDispatcher::handle_command(
 }
 
+void MDSRank::command_drop_cache(stringstream *ds)
+{
```
add `Mutex::Locker l(mds_lock);`

rishabh-d-dave replied:
@batrick Using `Mutex::Locker l(mds_lock);` creates the same issue as before [1]. At least for time being, I am replacing it by `mds_lock.Lock()` so that MDS would stay locked until control reaches here [2] after exiting `command_drop_cache()`; this prevents MDS from crashing. 

[1] https://github.com/ceph/ceph/pull/21566#discussion_r184181266
[2] https://github.com/ceph/ceph/blob/master/src/mds/MDSDaemon.cc#L1143

batrick replied:
Oh, right that lock is not necessary because it's locked in `MDSDaemon::ms_dispatch`.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -2983,11 +2983,71 @@ bool MDSRankDispatcher::handle_command(
+
+  // Optionally, flush the journal -
+  Formatter *fmtr = new JSONFormatter(true);
+  command_flush_journal(fmtr);
```
should call `_command_flush_journal`.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -2983,11 +2983,71 @@ bool MDSRankDispatcher::handle_command(
+	    "CEPH_SESSION_RECALL_STATEMSG_ACK from all sessions." << dendl;
+
+  // Optionally, flush the journal -
+  Formatter *fmtr = new JSONFormatter(true);
```
Use:

`std::unique_ptr<Formatter> fmtr(new JSONFormatter(true));`

then

`fmtr.reset(new JSONFormatter(true));`

when you want to create a new one.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -2983,11 +2983,71 @@ bool MDSRankDispatcher::handle_command(
+  if (r != 0)
+   *ds << "Failed to get cache status: " << cpp_strerror(r);
+  fmtr->flush(*ds);
+}
```
As I noted elsewhere, what you should actually do is concatenate all of the commands outputs into one JSON object. So you should be using the same `JSONFormatter` for all of these commands. Just create a new object section before calling each command.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -2983,11 +2983,71 @@ bool MDSRankDispatcher::handle_command(
+
+    time_passed = mono_clock::now() - time_at_beg;
+    if (TIMEOUT - time_passed.count() > 2)
+      sleep(2);
```
This needs to be using waiter/gather continuations. @ukernel can you suggest the right approach.

ukernel replied:
see Server::flush_client_sessions()

rishabh-d-dave replied:
I originally wrote this code to make MDS consume less CPU. Should it be gotten rid of? I am confused since discussion of waiter/gather continuations began from here.

rishabh-d-dave replied:
Done.

I was under the impression that `sleep` would lead to lesser number of iterations (when ACKs are not received spontaneously) allowing other threads/processes to use that CPU  time instead.

batrick commented on src/mds/Server.cc:
```diff
@@ -465,6 +465,11 @@ void Server::handle_client_session(MClientSession *m)
 
+  case CEPH_SESSION_RECALL_STATEMSG_ACK:
+    if (this->report_acks)
+      session->recall_acked = true;
```
The right way to do this is to record a sequence number for the recall messages the MDS sends so that the client can ACK the seq number. Then we know which recall messages it has handled. Recall what Zheng said in the email thread: "we can add a recall_state seq to MClientCapRelease and MClientSession message.".

Please modify those messages (and increment the message version) to include the sequence number.

rishabh-d-dave replied:
Please correct me if I am wrong: in my understanding, I need to assign seq. numbers to `CEPH_SESSION_RECALL_STATE` messages being sent by MDS in `Server::client_recall_state`, store them for later and then while receiving ACKs (until `mds_session_recall_ack_timeout`) for these messages I need to check them and report the ones (in the logs) I haven't received an ACK for.

> Recall what Zheng said in the email thread: "we can add a recall_state seq to MClientCapRelease and MClientSession message.".

Yeah, I couldn't comprehend the "seq" part back then.

rishabh-d-dave replied:
Updating the PR as per my understanding above.

batrick commented on src/mds/Server.h:
```diff
@@ -91,6 +93,7 @@ class Server {
 
 public:
   bool terminating_sessions;
+  set<int> crs_msg_seq_nums;
```
This should actually be a pair of sequence numbers (`version_t`) that are associated with each session (just like seq numbers in TCP). The first would be what the MDS sends and sequentially increases with each new `CEPH_SESSION_RECALL_STATE` message. The second would be what the client sends back as finished, via the `CEPH_SESSION_RECALL_STATEMSG_ACK` message.

But now that I'm reading the code more in SessionMap.h, I'm wondering if we can reuse the existing `CEPH_SESSION_FLUSHMSG` code for this purpose. There is already a sequence number in the session to use (see the code for `Session::cap_push_seq`. @ukernel would this be appropriate or should we use a separate sequence number?

ukernel replied:
using Session::cap_push_seq is appropriate. 

If client always sends ACK immediately after receiving drop cache request, existing Session::{wait_for_flush,finish_flush} code can serve this purpose completely. If not, need introduce a new waiter list ( similar to Session::waitfor_flush)

rishabh-d-dave replied:
@batrick 

> This should actually be a pair of sequence numbers (version_t) that are associated with each session (just like seq numbers in TCP). The first would be what the MDS sends and sequentially increases with each new CEPH_SESSION_RECALL_STATE message. The second would be what the client sends back as finished, via the CEPH_SESSION_RECALL_STATEMSG_ACK message.

If sequence number of the message sent by MDS is different from that of the client, I wouldn't know which sessions have not been trimmed and it will not be possible to report those sessions in the log. Though I can still see how many sessions have not trimmed. Will that be okay?

ukernel replied:
check code that handle CEPH_SESSION_FLUSHMSG/CEPH_SESSION_FLUSHMSG_ACK. they use the same seq number

rishabh-d-dave replied:
@ukernel Done. Thanks.  :)

ukernel commented on src/client/Client.cc:
```diff
@@ -2078,6 +2078,9 @@ void Client::handle_client_session(MClientSession *m)
     trim_caps(session, m->get_max_caps());
+    session->con->send_message(new MClientSession(
+				CEPH_SESSION_RECALL_STATEMSG_ACK,
+				m->get_seq()));
```
Does client always send ACK immediately

rishabh-d-dave replied:
@ukernel Client will always send ACKs but I don't if its fast enough to be immediate as it calls trim_caps() first.

rishabh-d-dave commented:
Should the command `drop cache` be changed to `cache drop`? There are two more cache-related commands -  `cache status` and `dump cache`- but they do not make the preferred order clear. I think `cache` preceding the operation is better.

batrick commented:
`cache drop` sounds good to me.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -3054,11 +3054,77 @@ bool MDSRankDispatcher::handle_command(
+  }
+
+  if (not mdcache->trim(UINT64_MAX))
+    return;
```
I think trimming the cache should occur after we get all the client recall acks. We want to trim the MDS cache when the client caps are released.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -3054,11 +3054,77 @@ bool MDSRankDispatcher::handle_command(
+  fmtr.get()->open_object_section("result");
+
+  int retval = _command_flush_journal(&ss);
+  fmtr.get()->open_object_section("flush_journal");
```
can be `fmtr->open_object_section...`

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -3054,11 +3054,77 @@ bool MDSRankDispatcher::handle_command(
+
+  sessionmap.get_client_session_set(unacked_sessions);
+
+  while (time_passed.count() < TIMEOUT) {
```
The MDS should do a wait, not an infinite loop. You need to add a waiter interface for this.

rishabh-d-dave replied:
> The MDS should do a wait, not an infinite loop.

It's not infinite. The looping stops on timeout.

batrick replied:
Sorry my wording was wrong. It's not infinite but it **is** a busy-wait. i.e. the CPU loops until the condition is true... for `TIMEOUT` seconds. This isn't acceptable.

rishabh-d-dave replied:
Got it.

batrick commented on src/include/ceph_fs.h:
```diff
@@ -301,6 +301,7 @@ enum {
 	CEPH_SESSION_RENEWCAPS,
 	CEPH_SESSION_STALE,
 	CEPH_SESSION_RECALL_STATE,
+	CEPH_SESSION_RECALL_STATEMSG_ACK,
```
this needs to go at the end of the enum otherwise the binary compatibility of the state messages is compromised

batrick commented on src/mds/Server.cc:
```diff
@@ -466,6 +466,12 @@ void Server::handle_client_session(MClientSession *m)
+  case CEPH_SESSION_RECALL_STATEMSG_ACK:
+    finish_flush_session(session, m->get_seq());
+    if (this->report_acks)
+      session->recall_acked = true;
```
`recall_acked = m->get_seq()`

batrick commented on src/mds/Server.cc:
```diff
@@ -1122,8 +1128,11 @@ void Server::recover_filelocks(CInode *in, bufferlist locks, int64_t client)
+void Server::recall_client_state(double ratio, bool report_acks)
 {
+  MClientSession *m;
+  static MDSGatherBuilder gather(g_ceph_context);
```
why static?

rishabh-d-dave replied:
I saw MDS crash without it -

```
(gdb)
1155      set<Session*> sessions;
(gdb)
1133      MDSGatherBuilder gather(g_ceph_context);
(gdb)

Program received signal SIGABRT, Aborted.
0x00007f7fd283b277 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56        return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
(gdb)
handle_fatal_signal (signum=6)
    at /home/centos/repos/ceph/src/global/signal_handler.cc:92
92      {
(gdb) c
Continuing.
```
I do not have a higher understanding of how MDS runs as a daemon but I saw the line being executed multiple times. So, the making it `static` seemed like a natural solution.

batrick commented on src/mds/Server.cc:
```diff
@@ -1122,8 +1128,11 @@ void Server::recover_filelocks(CInode *in, bufferlist locks, int64_t client)
-void Server::recall_client_state(void)
+void Server::recall_client_state(double ratio, bool report_acks)
 {
+  MClientSession *m;
```
Why did you move this var?

rishabh-d-dave replied:
Moved it during re-writings. Undid this change.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -3054,11 +3054,77 @@ bool MDSRankDispatcher::handle_command(
+  if (not mdcache->trim(UINT64_MAX))
+    return;
+
+  server->recall_client_state(1, true);
```
pass a pointer to

`MDSGatherBuilder gather(g_ceph_context);`

to `server->recall_client_state` which adds to the gather:

https://github.com/ceph/ceph/pull/21566/files#diff-eecf840e09d7481f7b8f415789852c7dR1177

then you can do a `gather.activate()` in `_wait_for_client_recall_state_acks` instead of the cpu-intensive loop.

rishabh-d-dave replied:
I can't figure out how to replace the busy/wait loop by `gather.activate()`. Also, calling `gather.activate()` in `_wait_for_client_recall_state_acks` leads to a crash due to the assert in `gather.activate()` [1]

And would these changes mean no need timeout? If so, the option `mds_session_recall_ack_timeout` will be redundant.

[1] https://github.com/ceph/ceph/blob/master/src/include/Context.h#L476

batrick commented:
@rishabh-d-dave status on this?

rishabh-d-dave commented:
Copying from here - https://github.com/ceph/ceph/pull/21566#discussion_r196117923.
This was in context to the comment above it - https://github.com/ceph/ceph/pull/21566#discussion_r194561415

@batrick I can't figure out how to replace the busy/wait loop by gather.activate(). Also, calling gather.activate() in _wait_for_client_recall_state_acks leads to a crash due to the assert in gather.activate() [1]

And would these changes mean no need timeout? If so, the option mds_session_recall_ack_timeout will be redundant.

[1] https://github.com/ceph/ceph/blob/master/src/include/Context.h#L476

batrick commented:
@rishabh-d-dave I added a commit to this PR which uses the gather interface. It works but you will want to add the timeout handling back in. I'm not yet sure yet on the right way to do that. Let's discuss in standup tomorrow. Feel free to squash the commits (i.e. drop my commit) and make adjustments to the code.

rishabh-d-dave commented:
The reason why ACKs are not received in time might be [this](https://github.com/rishabh-d-dave/ceph/blob/master/src/include/Context.h#L349) lock.

rishabh-d-dave commented on src/mds/MDSContext.h:
```diff
@@ -187,6 +188,15 @@ class MDSGather : public C_GatherBase<MDSInternalContextBase, MDSInternalContext
+  MDSContextSleeper(const std::string &name, MDSRank *mds) :
+    MDSInternalContext(mds), C_SaferCond(name) {}
+  void finish(int r) {
+    return;
```
Should I implement this? Do we have an example?

By the way, just in case if this comes in handy, ACKs are received by the time control reaches this line; observed during a gdb session.

batrick commented on src/client/Client.cc:
```diff
@@ -2083,6 +2083,9 @@ void Client::handle_client_session(MClientSession *m)
   case CEPH_SESSION_RECALL_STATE:
     trim_caps(session, m->get_max_caps());
+    session->con->
+    send_message(new MClientSession(CEPH_SESSION_RECALL_STATEMSG_ACK,
```
These messages are not necessary which is why I removed them in the previous version. You can just use the flush messages.

rishabh-d-dave replied:
Okay.

What about [session->crs_ack_rcvd](https://github.com/ceph/ceph/pull/21566/files#diff-23bc98a965757649a7e2d936e1eb7092R104)? Should it stay and where should the [code that checks if all ACKs are received](https://github.com/ceph/ceph/pull/21566/files#diff-ea4df8dd62d8f79137676b9f21d58633R3144) be (not just in textual region but also while execution)?

batrick replied:
You don't need that but you would want to keep track of which sessions have not dropped caps yet (indicated by which sessions have not sent back `flushmsg_ack` by your timeout). The `Gather` mechanism doesn't have a way to do that. So, you will need to create a subclass of `MDSContextSleeper` which keeps a `std::set<Session *>`. You will also need to virtualize `C_GatherBase::new_sub` and `C_GatherBase::sub_finish`` so you can override these methods in order to track the completions (you will still call `C_GatherBase::new_sub()` in your the overriden method to do the bulk of the work).

Also, consider allowing specifying a timeout via the command argument to `ceph tell cache drop` rather than config option.

rishabh-d-dave commented:
Usually when caps.size() is greater than newlim[1], the ACKs are not sent. However, I tweak the code to reverse this condition[1] to force ACKs while testing. Running `cache drop` went fine for the first time but the for rest of the time it hangs.

[1] https://github.com/rishabh-d-dave/ceph/blob/mds-drop-cache-command/src/mds/Server.cc#L1166

batrick commented:
> Usually when caps.size() is greater than newlim[1], the ACKs are not sent. However, I tweak the code to reverse this condition[1] to force ACKs while testing. Running cache drop went fine for the first time but the for rest of the time it hangs.

I think it's okay to skip sessions with fewer than the limit. For your testing, you should just create a lot of files `cp -av /usr .` to test.

rishabh-d-dave commented on src/mds/SessionMap.h:
```diff
@@ -666,5 +665,29 @@ class SessionMap : public SessionMapStore {
+    return new_sub();
+  }
+
+  ContextType *new_sub() override {
```
Apparently, `ContextType` is not visible in this file; the code fails to compile. Neither greping over the codebase revealed its definition[1] nor searching on web lead to any clues about it (in case it is defined is standard or boost libraries).

I coud use `MDSInternalContextBase` instead but then these function members would not be overriden.

[1] https://paste.fedoraproject.org/paste/Au5ZsTJghrN9A~c~YkOTYw

rishabh-d-dave commented:
>> **rishabh-d-dave 6 days ago**
>> Okay.
>>
>> What about session->crs_ack_rcvd? Should it stay and where should the code that checks if all ACKs are received be (not just in textual region but also while execution)?
>
>  **batrick 4 days ago Member**
>
> You don't need that but you would want to keep track of which sessions have not dropped caps yet (indicated by which sessions have not sent back flushmsg_ack by your timeout).
 
By "that" do you mean `session->crs_ack_rcvd`? I get that `CEPH_SESSION_RECALL_STATEMSG_ACK` is redundant, but I don't know how to keep a track of sessions that have/haven't received `flushmsg_ack` otherwise.
 
> The Gather mechanism doesn't have a way to do that. So, you will need to create a subclass of MDSContextSleeper which keeps a std::set<Session *>. You will also need to virtualize C_GatherBase::new_sub and C_GatherBase::sub_finish`` so you can override these methods in order to track the completions (you will still callC_GatherBase::new_sub()` in your the overriden method to do the bulk of the work).
 
Moved `C_GatherBase::sub_finish` from `private` to `protected`, though I think `protected` is not aboslutely correct since, except for the new case here, it should not be called by any derived class's instances.

I've added `has_subs()` to `C_GatherBase`. Are you sure that I need to virtualize `C_GatherBase`'s member function and not `C_GatherBuilderBase`'s? `MDSGatherBuilder` seemed to use `C_GatherBuilderBase` since its object had the member `has_subs()`.

> Also, consider allowing specifying a timeout via the command argument to ceph tell cache drop rather than config option.

I haven't implemented it yet, but I will.

batrick commented:
> By "that" do you mean session->crs_ack_rcvd?

Yes

> I get that CEPH_SESSION_RECALL_STATEMSG_ACK is redundant, but I don't know how to keep a track of sessions that have/haven't received flushmsg_ack otherwise.

You can keep track by subclassing MDSContextSleeper and tracking creations of "subs", i.e. what you quoted next:

>> The Gather mechanism doesn't have a way to do that. So, you will need to create a subclass of MDSContextSleeper which keeps a std::set<Session *>. You will also need to virtualize C_GatherBase::new_sub and C_GatherBase::sub_finish`` so you can override these methods in order to track the completions (you will still callC_GatherBase::new_sub()` in your the overriden method to do the bulk of the work).

>Moved C_GatherBase::sub_finish from private to protected, though I think protected is not aboslutely correct since, except for the new case here, it should not be called by any derived class's instances.

protected is okay IMO.

> I've added has_subs() to C_GatherBase.

No that needs to stay in GatherBaseBuilder.

> Are you sure that I need to virtualize C_GatherBase's member function and not C_GatherBuilderBase's? MDSGatherBuilder seemed to use C_GatherBuilderBase since its object had the member has_subs().

I mispooke in the earlier quote where I said you need to subclass MDSContextSleeper. I meant MDSGather. It will look like:

````
class MDSGatherSessions : public MDSGather {
  std::set<Session *> sessions;
  MDSInternalContextBase *new_sub(Session *s) {
    sessions.insert(s);
    return MDSGather::new_sub();
  }
}
typedef C_GatherBuilderBase<MDSInternalContextBase, MDSGatherSessions> MDSGatherSessionsBuilder;
````

Similar code for `finish_sub`. Then you can check which sessions have finished and which have not. You will need to modify `recall_client_state` to take `MDSGatherSessionsBuilder` instead of `MDSGatherBuilder`.

rishabh-d-dave commented:
>> Are you sure that I need to virtualize C_GatherBase's member function and not C_GatherBuilderBase's? MDSGatherBuilder seemed to use C_GatherBuilderBase since its object had the member has_subs().
>
>I mispooke in the earlier quote where I said you need to subclass MDSContextSleeper. I meant MDSGather. It will look like:
>
>````
>class MDSGatherSessions : public MDSGather {
>  std::set<Session *> sessions;
>  MDSInternalContextBase *new_sub(Session *s) {
>    sessions.insert(s);
>    return MDSGather::new_sub();
>  }
>}
>typedef C_GatherBuilderBase<MDSInternalContextBase, MDSGatherSessions> MDSGatherSessionsBuilder;
>````
>
>Similar code for `finish_sub`. Then you can check which sessions have finished and which have not. You will need to modify `recall_client_state` to take `MDSGatherSessionsBuilder` instead of `MDSGatherBuilder`.

Okay. I had written (in previous PR update) a class looking much like MDSGather, I've re-written it to be MDSGatherSessions. But now the issue is that the following line from [here](https://github.com/ceph/ceph/pull/21566/files#diff-ea4df8dd62d8f79137676b9f21d58633R3099) creates `gather` with `C_GatherBaseBuilder` type, so the methods `new_sub()` and `sub_finish()` in `C_GatherBase` won't be overriden -

```
    MDSGatherSessionsBuilder gather(g_ceph_context, sleeper);

```

Replacing `MDSGatherSessionsBuilder` by `MDSGatherSessions` would make `gather` an object of `C_GatherBase` but then `has_subs()` would be missing. So, either we add `has_subs()` to `C_GatherBase` or virtualize and override `new_sub()` in `C_GatherBuilderBase` I've tried latter but due to an issue I report below I couldn't actually check.

I get the following error g++ error for [this line](https://github.com/ceph/ceph/pull/21566/files#diff-eecf840e09d7481f7b8f415789852c7dR1172) -

```
error: no matching function for call to ‘C_GatherBuilderBase<MDSInternalContextBase, MDSGatherSessions>::new_sub(Session*&)’
         version_t seq = session->wait_for_flush(gather->new_sub(session));

```

I've tried changing parameter type in the `new_sub` but I couldn't get rid of the error.

I circumvented the compilation error by calling super-classes's `new_sub` without passing or inserting `session` to the set to check at least if overriding is fine, but the override seems ineffective -

```
Breakpoint 1, Server::recall_client_state (this=0x560441df8b60, ratio=ratio@entry=1, gather=gather@entry=0x7f35841421b0) at /home/user/ceph/src/mds/Server.cc:1171
1171	      if (gather && session->connection->has_feature(CEPH_FEATURE_EXPORT_PEER)) {
(gdb) n
1172	        version_t seq = session->wait_for_flush(gather->new_sub());
(gdb) s
new_sub (this=0x7f35841421b0) at /home/user/ceph/src/include/Context.h:468
468	    if (!c_gather) {
(gdb) n
Server::recall_client_state (this=0x560441df8b60, ratio=ratio@entry=1, gather=gather@entry=0x7f35841421b0) at /home/user/ceph/src/mds/Server.cc:1172
1172	        version_t seq = session->wait_for_flush(gather->new_sub());
```

I suppose it's because the `new_sub`'s signature doesn't match across sub- and super-class. But if substitute the template definition by <MDSInternalContextBase, MDSInternalContextGather>, it seems overriding should've worked.

```
// From C_GatherBase
  virtual ContextType *new_sub() {
// From C_GatherBaseBuilder
  ContextType *new_sub() {
// From MDSGatherSessions
  MDSInternalContextBase *new_sub() override {
```

batrick commented:
@rishabh-d-dave I'm working on a fix for this. It's nearly done but I have a small bug to figure out. I'll let you know how it's going tomorrow.

batrick commented:
Dependency on: https://github.com/ceph/ceph/pull/23134

batrick commented:
@rishabh-d-dave it now works:

````
$ bin/ceph daemon mds.a cache drop
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{
    "flush_journal": {
        "message": "",
        "return_code": 0
    },
    "cache": {
        "pool": {
            "items": 169,
            "bytes": 48936
        }
    }
}
````

and from the log:

````
2018-07-19 17:41:43.418 7fad9f9ba700  1 mds.a asok_command: cache drop (starting...)
2018-07-19 17:41:43.418 7fad9f9ba700  0 mds.0.server max_caps_per_client 0 < min_caps_per_client 100
2018-07-19 17:41:43.418 7fad9f9ba700 10 mds.0.server recall_client_state 1, caps per client 100-101
2018-07-19 17:41:43.418 7fad9f9ba700  0 mds.0.4  Sleeper r is 0
2018-07-19 17:41:43.418 7fad9f9ba700  7 mds.0.cache trim bytes_used=47kB limit=1GB reservation=0.05% count=18446744073709551615
2018-07-19 17:41:43.418 7fad9f9ba700  7 mds.0.cache trim_lru trimming 18446744073709551615 items from LRU size=10 mid=0 pintail=0 pinned=10
2018-07-19 17:41:43.418 7fad9f9ba700  7 mds.0.cache trim_lru trimmed 0 items
2018-07-19 17:41:43.418 7fad9f9ba700  7 mds.0.log _prepare_new_segment seq 2
2018-07-19 17:41:43.418 7fad9f9ba700 10 mds.0.log Advancing to next stray directory on mds 0
2018-07-19 17:41:43.418 7fad9f9ba700  7 mds.0.log _journal_segment_subtree_map
2018-07-19 17:41:43.418 7fad9f9ba700 10 mds.0.cache create_subtree_map 2 subtrees, 2 fullauth
2018-07-19 17:41:43.418 7fad9f9ba700 15 mds.0.cache show_subtrees
2018-07-19 17:41:43.418 7fad9f9ba700 10 mds.0.cache |__ 0    auth [dir 0x1 / [2,head] auth v=1 cv=1/1 REP dir_auth=0 state=1073741825|complete f() n() hs=0+0,ss=0+0 | subtree=1 dirty=0 authpin=0 0x701fcc6d00]
2018-07-19 17:41:43.418 7fad9f9ba700 10 mds.0.cache |__ 0    auth [dir 0x100 ~mds0/ [2,head] auth v=1 cv=1/1 dir_auth=0 state=1073741825|complete f(v0 10=0+10) n(v0 10=0+10) hs=10+0,ss=0+0 | child=1 subtree=1 dirty=0 authpin=0 0x701fcc7200]
2018-07-19 17:41:43.418 7fad9f9ba700 15 mds.0.cache  subtree [dir 0x1 / [2,head] auth v=1 cv=1/1 REP dir_auth=0 state=1073741825|complete f() n() hs=0+0,ss=0+0 | subtree=1 dirty=0 authpin=0 0x701fcc6d00]
2018-07-19 17:41:43.418 7fad9f9ba700 15 mds.0.cache  subtree [dir 0x100 ~mds0/ [2,head] auth v=1 cv=1/1 dir_auth=0 state=1073741825|complete f(v0 10=0+10) n(v0 10=0+10) hs=10+0,ss=0+0 | child=1 subtree=1 dirty=0 authpin=0 0x701fcc7200]
2018-07-19 17:41:43.418 7fad9f9ba700 20 mds.0.journal EMetaBlob::add_dir_context final:
2018-07-19 17:41:43.418 7fad9f9ba700 20 mds.0.journal EMetaBlob::add_dir_context final:
2018-07-19 17:41:43.418 7fad9f9ba700 15 mds.0.cache  subtrees {0x1=[],0x100=[]}
2018-07-19 17:41:43.418 7fad9f9ba700 15 mds.0.cache  ambiguous_subtrees
2018-07-19 17:41:43.418 7fad969a8700  5 mds.0.log _submit_thread 4195196~872 : ESubtreeMap 2 subtrees , 0 ambiguous [metablob 0x1, 2 dirs]
2018-07-19 17:41:43.418 7fad969a8700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6836/17378 -- osd_op(unknown.0.4:40 2.4 2:2e2fa760:::200.00000001:head [write 892~892 [fadvise_dontneed]] snapc 0=[] ondisk+write+known_if_redirected+full_force e16) v8 -- 0x7020061a80 con 0
2018-07-19 17:41:43.418 7fad969a8700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6836/17378 -- osd_op(unknown.0.4:41 2.4 2:292cf221:::200.00000000:head [writefull 0~90 [fadvise_dontneed]] snapc 0=[] ondisk+write+known_if_redirected+full_force e16) v8 -- 0x702005e680 con 0
2018-07-19 17:41:43.434 7fada01bb700  1 -- 127.0.0.1:6846/1607993607 <== osd.1 127.0.0.1:6836/17378 31 ==== osd_op_reply(40 200.00000001 [write 892~892 [fadvise_dontneed]] v16'9 uv9 ondisk = 0) v8 ==== 156+0+0 (1456679394 0 0) 0x702005e9c0 con 0x701fff1800
2018-07-19 17:41:43.434 7fad971a9700 10 MDSIOContextBase::complete: 13C_MDL_Flushed
2018-07-19 17:41:43.434 7fad971a9700 10 MDSIOContextBase::complete: 13C_MDL_Flushed
2018-07-19 17:41:43.434 7fad971a9700 10 MDSInternalContextBase::complete: 25MDSInternalContextWrapper
2018-07-19 17:41:43.434 7fad971a9700 10 MDSIOContextBase::complete: 12C_IO_Wrapper
2018-07-19 17:41:43.434 7fad971a9700 10 MDSInternalContextBase::complete: 25MDSInternalContextWrapper
2018-07-19 17:41:43.434 7fad9f9ba700  5 mds.0.4 _command_flush_journal: beginning segment expiry
2018-07-19 17:41:43.434 7fad9f9ba700 10 mds.0.log trim_all: 2/0/0
2018-07-19 17:41:43.434 7fad9f9ba700 10 mds.0.openfiles commit log_seq 2
2018-07-19 17:41:43.434 7fad9f9ba700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6836/17378 -- osd_op(unknown.0.4:42 2.6 2:654134d2:::mds0_openfiles.0:head [omap-set-header] snapc 0=[] ondisk+write+known_if_redirected+full_force e16) v8 -- 0x70200f2000 con 0
2018-07-19 17:41:43.434 7fad9f9ba700  6 mds.0.journal LogSegment(1/4194304).try_to_expire
2018-07-19 17:41:43.434 7fad9f9ba700  4 mds.0.sessionmap save_if_dirty: writing 0
2018-07-19 17:41:43.434 7fad9f9ba700  6 mds.0.journal LogSegment(1/4194304).try_to_expire success
2018-07-19 17:41:43.434 7fad9f9ba700 10 mds.0.log try_expire expired segment 1/4194304
2018-07-19 17:41:43.434 7fad9f9ba700  5 mds.0.log _expired segment 1/4194304, 1 events
2018-07-19 17:41:43.434 7fad9f9ba700 10 mds.0.log _trim_expired_segments open file table committedseq 1 <= 1/4194304
2018-07-19 17:41:43.434 7fad9f9ba700  5 mds.0.4 _command_flush_journal: waiting for 0 segments to expire
2018-07-19 17:41:43.434 7fad9f9ba700  5 mds.0.4 _command_flush_journal: expiry complete, expire_pos/trim_pos is now 400000/400000
2018-07-19 17:41:43.434 7fad9f9ba700 10 mds.0.log _trim_expired_segments open file table committedseq 1 <= 1/400000
2018-07-19 17:41:43.434 7fad9f9ba700  5 mds.0.4 _command_flush_journal: trim complete, expire_pos/trim_pos is now 400000/400000
2018-07-19 17:41:43.434 7fad9f9ba700  1 -- 127.0.0.1:6846/5fd80907 --> 127.0.0.1:6836/43e2 -- osd_op(unknown.0.4:2b 2.4 2:292cf221:::200.00000000:head [writefull 0~90 [fadvise_dontneed]] snapc 0=[] ondisk+write+known_if_redirected+full_force e16) v8 -- 0x70200f2340 con 0
2018-07-19 17:41:43.442 7fada01bb700  1 -- 127.0.0.1:6846/1607993607 <== osd.1 127.0.0.1:6836/17378 32 ==== osd_op_reply(41 200.00000000 [writefull 0~90 [fadvise_dontneed]] v16'10 uv10 ondisk = 0) v8 ==== 156+0+0 (3736483682 0 0) 0x702005e680 con 0x701fff1800
2018-07-19 17:41:43.450 7fada01bb700  1 -- 127.0.0.1:6846/1607993607 <== osd.1 127.0.0.1:6836/17378 33 ==== osd_op_reply(43 200.00000000 [writefull 0~90 [fadvise_dontneed]] v16'11 uv11 ondisk = 0) v8 ==== 156+0+0 (2855701686 0 0) 0x702005e680 con 0x701fff1800
2018-07-19 17:41:43.450 7fada01bb700  1 -- 127.0.0.1:6846/1607993607 <== osd.1 127.0.0.1:6836/17378 34 ==== osd_op_reply(42 mds0_openfiles.0 [omap-set-header] v16'5 uv5 ondisk = 0) v8 ==== 160+0+0 (3953806500 0 0) 0x702005e680 con 0x701fff1800
2018-07-19 17:41:43.450 7fad971a9700 10 MDSIOContextBase::complete: 13C_IO_OFT_Save
2018-07-19 17:41:43.450 7fad971a9700 10 mds.0.openfiles _commit_finish log_seq 2
2018-07-19 17:41:43.450 7fad971a9700 10 MDSInternalContextBase::complete: 15C_OFT_Committed
2018-07-19 17:41:43.450 7fad971a9700 10 mds.0.log _trim_expired_segments trimming expired 1/0x400000
2018-07-19 17:41:43.450 7fad971a9700 10 mds.0.log _trim_expired_segments waiting for 2/4195196 to expire
2018-07-19 17:41:43.450 7fad971a9700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6836/17378 -- osd_op(unknown.0.4:44 2.4 2:292cf221:::200.00000000:head [writefull 0~90 [fadvise_dontneed]] snapc 0=[] ondisk+write+known_if_redirected+full_force e16) v8 -- 0x70200f2000 con 0
2018-07-19 17:41:43.450 7fad9f9ba700  5 mds.0.4 _command_flush_journal: write_head complete, all done!
2018-07-19 17:41:43.450 7fad9f9ba700  1 mds.a asok_command: cache drop (complete)
````

HOWEVER, if we're doing I/O from a client (`cp -av /usr .`), I've observed some form of deadlock:

````
2018-07-19 17:43:01.101 7fad9d1b5700  3 mds.0.server handle_client_session client_session(flushmsg_ack seq 55) v1 from client.4213
2018-07-19 17:43:01.101 7fad9d1b5700  7 mds.0.4 mds has 1 queued contexts
2018-07-19 17:43:01.101 7fad9d1b5700 10 mds.0.4 0x701fca06a0
2018-07-19 17:43:01.101 7fad9d1b5700 10 mds.0.4  finish 0x701fca06a0
2018-07-19 17:43:01.101 7fad9f9ba700  0 mds.0.4  Sleeper r is 0
2018-07-19 17:43:01.101 7fad9f9ba700  7 mds.0.cache trim bytes_used=1MB limit=1GB reservation=0.05% count=18446744073709551615
2018-07-19 17:43:01.101 7fad9f9ba700  7 mds.0.cache trim_lru trimming 18446744073709551615 items from LRU size=28 mid=0 pintail=10 pinned=28
2018-07-19 17:43:01.101 7fad9f9ba700  7 mds.0.cache trim_lru trimmed 0 items
2018-07-19 17:43:01.101 7fad9f9ba700  7 mds.0.log _prepare_new_segment seq 593
2018-07-19 17:43:01.101 7fad9f9ba700 10 mds.0.log Advancing to next stray directory on mds 0
2018-07-19 17:43:01.101 7fad9f9ba700  7 mds.0.log _journal_segment_subtree_map
2018-07-19 17:43:01.101 7fad9f9ba700 10 mds.0.cache create_subtree_map 2 subtrees, 2 fullauth
2018-07-19 17:43:01.101 7fad9f9ba700 15 mds.0.cache show_subtrees
2018-07-19 17:43:01.101 7fad9f9ba700 10 mds.0.cache |__ 0    auth [dir 0x1 / [2,head] auth pv=17 v=13 cv=1/1 REP dir_auth=0 ap=2+269+270 state=1610874881|complete f(v0 m2018-07-19 17:42:57.617196 1=0+1)->f(v0 m2018-07-19 17:42:57.617196 1=0+1) n(v0 rc2018-07-19 17:42:57.768478 8=4+4)/n()->n(v0 rc2018-07-19 17:43:00.263541 b9 185=146+39)/n() hs=1+0,ss=0+0 dirty=1 | child=1 subtree=1 dirty=1 authpin=1 0x701fcc6d00]
2018-07-19 17:43:01.101 7fad9f9ba700 10 mds.0.cache |__ 0    auth [dir 0x100 ~mds0/ [2,head] auth v=1 cv=1/1 dir_auth=0 state=1073741825|complete f(v0 10=0+10) n(v0 10=0+10) hs=10+0,ss=0+0 | child=1 subtree=1 dirty=0 authpin=0 0x701fcc7200]
2018-07-19 17:43:01.101 7fad9f9ba700 15 mds.0.cache  subtree [dir 0x1 / [2,head] auth pv=17 v=13 cv=1/1 REP dir_auth=0 ap=2+269+270 state=1610874881|complete f(v0 m2018-07-19 17:42:57.617196 1=0+1)->f(v0 m2018-07-19 17:42:57.617196 1=0+1) n(v0 rc2018-07-19 17:42:57.768478 8=4+4)/n()->n(v0 rc2018-07-19 17:43:00.263541 b9 185=146+39)/n() hs=1+0,ss=0+0 dirty=1 | child=1 subtree=1 dirty=1 authpin=1 0x701fcc6d00]
2018-07-19 17:43:01.101 7fad9f9ba700 15 mds.0.cache  subtree [dir 0x100 ~mds0/ [2,head] auth v=1 cv=1/1 dir_auth=0 state=1073741825|complete f(v0 10=0+10) n(v0 10=0+10) hs=10+0,ss=0+0 | child=1 subtree=1 dirty=0 authpin=0 0x701fcc7200]
2018-07-19 17:43:01.101 7fad9f9ba700 20 mds.0.journal EMetaBlob::add_dir_context final:
2018-07-19 17:43:01.101 7fad9f9ba700 20 mds.0.journal EMetaBlob::add_dir_context final:
2018-07-19 17:43:01.101 7fad9f9ba700 15 mds.0.cache  subtrees {0x1=[],0x100=[]}
2018-07-19 17:43:01.101 7fad9f9ba700 15 mds.0.cache  ambiguous_subtrees
2018-07-19 17:43:01.101 7fad969a8700  5 mds.0.log _submit_thread 5056869~872 : ESubtreeMap 2 subtrees , 0 ambiguous [metablob 0x1, 2 dirs]
2018-07-19 17:43:01.105 7fad969a8700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6836/17378 -- osd_op(unknown.0.4:50 2.4 2:2e2fa760:::200.00000001:head [write 75885~787572 [fadvise_dontneed]] snapc 0=[] ondisk+write+known_if_redirected+full_force e16) v8 -- 0x70200f3040 con 0
2018-07-19 17:43:01.105 7fad9d1b5700  1 -- 127.0.0.1:6846/1607993607 <== client.4213 127.0.0.1:0/2662882739 349 ==== client_request(client.4213:296 getattr pAsLsXsFs #0x1 2018-07-19 17:43:01.108858 caller_uid=0, caller_gid=0{0,}) v4 ==== 144+0+0 (1198522490 0 0) 0x70200f3040 con 0x7020066700
2018-07-19 17:43:01.105 7fad9d1b5700  4 mds.0.server handle_client_request client_request(client.4213:296 getattr pAsLsXsFs #0x1 2018-07-19 17:43:01.108858 caller_uid=0, caller_gid=0{0,}) v4
2018-07-19 17:43:01.105 7fad9d1b5700 20 mds.0.4 get_session have 0x70200cae00 client.4213 127.0.0.1:0/2662882739 state open
2018-07-19 17:43:01.105 7fad9d1b5700 15 mds.0.server  oldest_client_tid=27
2018-07-19 17:43:01.105 7fad9d1b5700  7 mds.0.cache request_start request(client.4213:296 cr=0x70200f3040)
2018-07-19 17:43:01.105 7fad9d1b5700  7 mds.0.server dispatch_client_request client_request(client.4213:296 getattr pAsLsXsFs #0x1 2018-07-19 17:43:01.108858 caller_uid=0, caller_gid=0{0,}) v4
2018-07-19 17:43:01.105 7fad9d1b5700 10 mds.0.server rdlock_path_pin_ref request(client.4213:296 cr=0x70200f3040) #0x1
2018-07-19 17:43:01.105 7fad9d1b5700  7 mds.0.cache traverse: opening base ino 0x1 snap head
2018-07-19 17:43:01.105 7fad9d1b5700 10  mds.0.cache.snaprealm(0x1 seq 1 0x701fecd900) have_past_parents_open [1,head]
2018-07-19 17:43:01.105 7fad9d1b5700 10 mds.0.cache path_traverse finish on snapid head
2018-07-19 17:43:01.105 7fad9d1b5700 10 mds.0.server ref is [inode 0x1 [...2,head] / auth v3 pv5 ap=269+0 snaprealm=0x701fecd900 dirtyparent f(v0 m2018-07-19 17:42:57.617196 1=0+1)->f(v0 m2018-07-19 17:42:57.617196 1=0+1) n(v0 1=0+1)->n(v0 1=0+1) (isnap sync r=269) (inest lock w=2 dirty) (ipolicy sync r=195) (ifile mix) (iversion lock w=2 last_client=4213) caps={4213=pAsLsXs/-@2} | dirtyscattered=1 request=0 lock=4 dirfrag=1 caps=1 dirtyparent=1 dirty=1 authpin=1 0x701fe7dc00]
2018-07-19 17:43:01.169 7fada01bb700  1 -- 127.0.0.1:6846/1607993607 <== osd.1 127.0.0.1:6836/17378 41 ==== osd_op_reply(50 200.00000001 [write 75885~787572 [fadvise_dontneed]] v16'17 uv17 ondisk = 0) v8 ==== 156+0+0 (241523181 0 0) 0x702101ed00 con 0x701fff1800
2018-07-19 17:43:01.169 7fad971a9700 10 MDSIOContextBase::complete: 18C_MDS_openc_finish
2018-07-19 17:43:03.109 7fad9a1af700 10 mds.beacon.a _send up:active seq 73
018-07-19 17:43:03.109 7fad9a1af700 10 monclient: _send_mon_message to mon.c at 127.0.0.1:40775/0
2018-07-19 17:43:03.109 7fad9a1af700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:40775/0 -- mdsbeacon(24164/a up:active seq 73 v5) v7 -- 0x701febd100 con 0
2018-07-19 17:43:03.985 7fad9b1b1700 20 mgrc operator() sending 24 counters (of possible 280), 0 new, 0 removed
2018-07-19 17:43:03.985 7fad9b1b1700 20 mgrc _send_report encoded 214 bytes
2018-07-19 17:43:03.985 7fad9b1b1700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6813/16760 -- mgrreport(unknown.a +0-0 packed 214) v6 -- 0x701febc300 con 0
2018-07-19 17:43:07.109 7fad9a1af700 10 mds.beacon.a _send up:active seq 74
2018-07-19 17:43:07.109 7fad9a1af700 10 monclient: _send_mon_message to mon.c at 127.0.0.1:40775/0
2018-07-19 17:43:07.109 7fad9a1af700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:40775/0 -- mdsbeacon(24164/a up:active seq 74 v5) v7 -- 0x701febd480 con 0
2018-07-19 17:43:07.969 7fad9c1b3700 10 monclient: tick
2018-07-19 17:43:07.969 7fad9c1b3700 10 monclient: _check_auth_rotating have uptodate secrets (they expire after 2018-07-19 17:42:37.970809)
2018-07-19 17:43:07.969 7fad9c1b3700 20 monclient: _un_backoff reopen_interval_multipler now 1
2018-07-19 17:43:08.985 7fad9b1b1700 20 mgrc operator() sending 24 counters (of possible 280), 0 new, 0 removed
2018-07-19 17:43:08.985 7fad9b1b1700 20 mgrc _send_report encoded 214 bytes
2018-07-19 17:43:08.985 7fad9b1b1700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6813/16760 -- mgrreport(unknown.a +0-0 packed 214) v6 -- 0x701febaa80 con 0
2018-07-19 17:43:11.109 7fad9a1af700 10 mds.beacon.a _send up:active seq 75
2018-07-19 17:43:11.109 7fad9a1af700 10 monclient: _send_mon_message to mon.c at 127.0.0.1:40775/0
2018-07-19 17:43:11.109 7fad9a1af700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:40775/0 -- mdsbeacon(24164/a up:active seq 75 v5) v7 -- 0x70200ccd80 con 0
2018-07-19 17:43:13.984 7fad9b1b1700 20 mgrc operator() sending 24 counters (of possible 280), 0 new, 0 removed
2018-07-19 17:43:13.984 7fad9b1b1700 20 mgrc _send_report encoded 214 bytes
2018-07-19 17:43:13.984 7fad9b1b1700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6813/16760 -- mgrreport(unknown.a +0-0 packed 214) v6 -- 0x701febb500 con 0
2018-07-19 17:43:15.108 7fad9a1af700 10 mds.beacon.a _send up:active seq 76
2018-07-19 17:43:15.108 7fad9a1af700 10 monclient: _send_mon_message to mon.c at 127.0.0.1:40775/0
2018-07-19 17:43:15.108 7fad9a1af700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:40775/0 -- mdsbeacon(24164/a up:active seq 76 v5) v7 -- 0x70200cb880 con 0
2018-07-19 17:43:17.968 7fad9c1b3700 10 monclient: tick
2018-07-19 17:43:17.968 7fad9c1b3700 10 monclient: _check_auth_rotating have uptodate secrets (they expire after 2018-07-19 17:42:47.971074)
2018-07-19 17:43:17.968 7fad9c1b3700 20 monclient: _un_backoff reopen_interval_multipler now 1
2018-07-19 17:43:17.968 7fad9f1b9700  1 heartbeat_map is_healthy 'MDSRank' had timed out after 15
2018-07-19 17:43:18.984 7fad9b1b1700 20 mgrc operator() sending 24 counters (of possible 280), 0 new, 0 removed
2018-07-19 17:43:18.984 7fad9b1b1700 20 mgrc _send_report encoded 214 bytes
2018-07-19 17:43:18.984 7fad9b1b1700  1 -- 127.0.0.1:6846/1607993607 --> 127.0.0.1:6813/16760 -- mgrreport(unknown.a +0-0 packed 214) v6 -- 0x701febb880 con 0
2018-07-19 17:43:19.108 7fad9a1af700  1 heartbeat_map is_healthy 'MDSRank' had timed out after 15
2018-07-19 17:43:19.108 7fad9a1af700  1 mds.beacon.a _send skipping beacon, heartbeat map not healthy
2018-07-19 17:43:22.968 7fad9f1b9700  1 heartbeat_map is_healthy 'MDSRank' had timed out after 15
2018-07-19 17:43:23.108 7fad9a1af700  1 heartbeat_map is_healthy 'MDSRank' had timed out after 15
2018-07-19 17:43:23.108 7fad9a1af700  1 mds.beacon.a _send skipping beacon, heartbeat map not healthy
````

It appears to be because we trimmed the cache then flushed the journal, causing some kind of problem.

````
$ sudo gdb -q -p 17862
Attaching to process 17862
[New LWP 17863]
[New LWP 17864]
[New LWP 17865]
[New LWP 17866]
[New LWP 17867]
[New LWP 17868]
[New LWP 17869]
[New LWP 17870]
[New LWP 17871]
[New LWP 17872]
[New LWP 17873]
[New LWP 17875]
[New LWP 17876]
[New LWP 17877]
[New LWP 18081]
[New LWP 18082]
[New LWP 18083]
[New LWP 18084]
[New LWP 18085]
[New LWP 18086]
[New LWP 18087]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
185     ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S: No such file or directory.
(gdb) info threads
  Id   Target Id         Frame 
* 1    Thread 0x7fadad9e6440 (LWP 17862) "ceph-mds" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  2    Thread 0x7fada115a700 (LWP 17863) "log" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  3    Thread 0x7fad9f1b9700 (LWP 17864) "service" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  4    Thread 0x7fad9f9ba700 (LWP 17865) "admin_socket" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  5    Thread 0x7fada01bb700 (LWP 17866) "msgr-worker-0" 0x00007fada3980a13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  6    Thread 0x7fad9e9b8700 (LWP 17867) "msgr-worker-1" 0x00007fada3980a13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  7    Thread 0x7fad9e1b7700 (LWP 17868) "msgr-worker-2" 0x00007fada3980a13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  8    Thread 0x7fad9d9b6700 (LWP 17869) "signal_handler" 0x00007fada397474d in poll () at ../sysdeps/unix/syscall-template.S:84
  9    Thread 0x7fad9d1b5700 (LWP 17870) "ms_dispatch" 0x00007fada420aa44 in std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
  10   Thread 0x7fad9c9b4700 (LWP 17871) "ms_local" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  11   Thread 0x7fad9c1b3700 (LWP 17872) "safe_timer" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  12   Thread 0x7fad9b9b2700 (LWP 17873) "fn_anonymous" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  13   Thread 0x7fad9b1b1700 (LWP 17875) "safe_timer" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  14   Thread 0x7fad9a9b0700 (LWP 17876) "safe_timer" __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
  15   Thread 0x7fad9a1af700 (LWP 17877) "safe_timer" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  16   Thread 0x7fad999ae700 (LWP 18081) "ms_dispatch" pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  17   Thread 0x7fad991ad700 (LWP 18082) "OpHistorySvc" 0x00007fada394530d in nanosleep () at ../sysdeps/unix/syscall-template.S:84
  18   Thread 0x7fad989ac700 (LWP 18083) "mds_rank_progr" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  19   Thread 0x7fad981ab700 (LWP 18084) "PQ_Finisher" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  20   Thread 0x7fad979aa700 (LWP 18085) "safe_timer" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  21   Thread 0x7fad971a9700 (LWP 18086) "fn_anonymous" __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
  22   Thread 0x7fad969a8700 (LWP 18087) "md_submit" pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
(gdb) thread 9
[Switching to thread 9 (Thread 0x7fad9d1b5700 (LWP 17870))]
#0  0x00007fada420aa44 in std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007fada420aa44 in std::_Rb_tree_insert_and_rebalance(bool, std::_Rb_tree_node_base*, std::_Rb_tree_node_base*, std::_Rb_tree_node_base&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x000000701df005fb in std::_Rb_tree<SimpleLock*, SimpleLock*, std::_Identity<SimpleLock*>, std::less<SimpleLock*>, std::allocator<SimpleLock*> >::_M_insert_<SimpleLock* const&, std::_Rb_tree<SimpleLock*, SimpleLock*, std::_Identity<SimpleLock*>, std::less<SimpleLock*>, std::allocator<SimpleLock*> >::_Alloc_node> (this=0x7fad9d1b08a0, __x=<optimized out>, __p=0x7021005db0, __v=@0x7fad9d1b0890: 0x701fe7e148, 
    __node_gen=...) at /usr/include/c++/7/bits/stl_tree.h:1755
#2  0x000000701df5dace in std::_Rb_tree<SimpleLock*, SimpleLock*, std::_Identity<SimpleLock*>, std::less<SimpleLock*>, std::allocator<SimpleLock*> >::_M_insert_unique<SimpleLock*>(SimpleLock*&&) (this=this@entry=0x7fad9d1b08a0, __v=__v@entry=<unknown type in /home/pdonnell/ceph/build/bin/ceph-mds, CU 0xe1a10f, DIE 0x120b1ab>) at /usr/include/c++/7/bits/stl_tree.h:2096
#3  0x000000701df1b0ed in std::set<SimpleLock*, std::less<SimpleLock*>, std::allocator<SimpleLock*> >::insert(SimpleLock*&&) (__x=<unknown type in /home/pdonnell/ceph/build/bin/ceph-mds, CU 0xe1a10f, DIE 0x12e19b2>, this=0x7fad9d1b08a0) at /usr/include/c++/7/bits/stl_set.h:510
#4  Server::handle_client_getattr (this=0x701fcded10, mdr=..., is_lookup=<optimized out>) at /home/pdonnell/ceph/src/mds/Server.cc:3093
#5  0x000000701df4b15b in Server::handle_client_request (this=this@entry=0x701fcded10, req=req@entry=0x70200f3040) at /home/pdonnell/ceph/src/mds/Server.cc:1771
#6  0x000000701df4ea8b in Server::dispatch (this=0x701fcded10, m=m@entry=0x70200f3040) at /home/pdonnell/ceph/src/mds/Server.cc:260
#7  0x000000701decab4c in MDSRank::handle_deferrable_message (this=this@entry=0x701ffac000, m=m@entry=0x70200f3040) at /home/pdonnell/ceph/src/mds/MDSRank.cc:767
#8  0x000000701deda89b in MDSRank::_dispatch (this=this@entry=0x701ffac000, m=m@entry=0x70200f3040, new_msg=new_msg@entry=true) at /home/pdonnell/ceph/src/mds/MDSRank.cc:604
#9  0x000000701dedaf95 in MDSRankDispatcher::ms_dispatch (this=0x701ffac000, m=m@entry=0x70200f3040) at /home/pdonnell/ceph/src/mds/MDSRank.cc:583
#10 0x000000701dec36d3 in MDSDaemon::ms_dispatch (this=0x701ff04000, m=0x70200f3040) at /home/pdonnell/ceph/src/mds/MDSDaemon.cc:1159
#11 0x00007fada4e95cca in Messenger::ms_deliver_dispatch (m=0x70200f3040, this=0x701fe78000) at /home/pdonnell/ceph/src/msg/Messenger.h:655
#12 DispatchQueue::entry (this=0x701fe78180) at /home/pdonnell/ceph/src/msg/DispatchQueue.cc:201
#13 0x00007fada4f3712d in DispatchQueue::DispatchThread::entry (this=<optimized out>) at /home/pdonnell/ceph/src/msg/DispatchQueue.h:102
#14 0x00007fada46f66ba in start_thread (arg=0x7fad9d1b5700) at pthread_create.c:333
#15 0x00007fada398041d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
````

It doesn't happen if you just use the `flush journal` command. @ukernel any idea? This is pretty trivial to reproduce in my testing.

rishabh-d-dave commented:
> rishabh-d-dave wote:
>> rishabh-d-dave wote:
>>
>>
>>  Also, consider allowing specifying a timeout via the command argument to ceph tell cache drop rather than config option.
>
> I haven't implemented it yet, but I will.
>
I've removed the conf file option for timeout and added the code to pass it to the command instead.

While testing the PR, I saw that not all the cache would be dropped at once. Instead, it happened gradually with every repetition of `cache drop` until MDS become unresponsive (i.e. "active (laggy)"). The console output is copied in the attached file[1].

[1] [cache-drop-repeatedly.log](https://github.com/ceph/ceph/files/2224412/cache-drop-repeatedly.log)

rishabh-d-dave commented:
@batrick @vshankar 
I had the following merge conflict on this PR -

```
  /// Wait until the \c secs expires or \c complete() is called
  int wait_for(double secs) {
    utime_t interval;
    interval.set_from_double(secs);
    Mutex::Locker l{lock};
    if (done) {
<<<<<<< HEAD
      return rval;
=======
      return 0;
>>>>>>> wip
    }
    cond.WaitInterval(lock, interval);
    return done ? rval : ETIMEDOUT;
  }
```

I kept `return rval` as I find it more correct. Both of these changes where made by Patrick in `7090ea24dd5` and `c11389cfc7c` respectively. I am reporting this because if `return 0` served some specific purpose or was some sort of hack, I would be wrong.

`git blame src/common/Cond.h` on master -

```
7090ea24dd5 src/common/Cond.h                 (Patrick Donnelly 2018-07-19 13:54:12 -0700 205)       return rval;
```

Corresponding code snippet -

```
  /// Wait until the \c secs expires or \c complete() is called
  int wait_for(double secs) {
    utime_t interval;
    interval.set_from_double(secs);
    Mutex::Locker l{lock};
    if (done) {
      return rval;
    }
    cond.WaitInterval(lock, interval);
    return done ? rval : ETIMEDOUT;
  }
```

`git blame src/common/Cond.h` on mds-drop-cache-command -

```
c11389cfc7c src/common/Cond.h                 (Patrick Donnelly 2018-07-19 13:36:52 -0700 207)       return 0;
```

Corresponding code snippet - 

```
  /// Wait until the \c secs expires or \c complete() is called
  int wait_for(double secs) {
    utime_t interval;
    interval.set_from_double(secs);
    Mutex::Locker l{lock};
    if (done) {
      return 0;
    }
    cond.WaitInterval(lock, interval);
    return done ? rval : ETIMEDOUT;
  }
``

vshankar commented:
> I kept return rval as I find it more correct. Both of these changes where made by Patrick in 7090ea24dd5 and c11389cfc7c respectively. I am reporting this because if return 0 served some specific purpose or was some sort of hack, I would be wrong.

you'd need to return `rval`. btw, where did the `return 0` in the merge conflict come from? was the change done locally and then rebased?

rishabh-d-dave commented:
>>    I kept return rval as I find it more correct. Both of these changes where made by Patrick in 7090ea2 and c11389c respectively. I am reporting this because if return 0 served some specific purpose or was some sort of hack, I would be wrong.
>
> you'd need to return rval. btw, where did the return 0 in the merge conflict come from? was the change done locally and then rebased?

`return 0` was part of the second commit on this PR before the merge conflict.

rishabh-d-dave commented on src/mds/MDSContext.h:
```diff
@@ -226,6 +227,45 @@ class MDSGather : public C_GatherBase<MDSInternalContextBase, MDSInternalContext
+    detached = true;
+  }
+private:
+  using locker = std::unique_lock<std::mutex>;
```
This part of the code was written by @batrick. I think he can answer better; I am not yet good with waiter-gather continuations and locks.

rishabh-d-dave commented on src/mds/MDSRank.cc:
```diff
@@ -2124,9 +2124,14 @@ bool MDSRankDispatcher::handle_asok_command(std::string_view command,
+    mdcache->cache_status(f);
+  } else if (command == "cache drop") {
+    int64_t timeout;
+    cmd_getval(g_ceph_context, cmdmap, "timeout", timeout);
```
Nope -
```
$ ./bin/ceph daemon mds.a cache drop 30
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{
    "flush_journal": {
        "message": "",
        "return_code": 0
    },
    "cache": {
        "pool": {
            "items": 730,
            "bytes": 68448
        }
    }
}
$ grep -rnI loglog out/
out/mds.a.log:5368:2018-08-08 06:53:25.734 7f8e12629700  0 mds.0.3 loglog, timeout = 30
$ ./bin/ceph daemon mds.a cache drop 45
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
{
    "flush_journal": {
        "message": "",
        "return_code": 0
    },
    "cache": {
        "pool": {
            "items": 730,
            "bytes": 68448
        }
    }
}
$ grep -rnI loglog out/
out/mds.a.log:5368:2018-08-08 06:53:25.734 7f8e12629700  0 mds.0.3 loglog, timeout = 30
out/mds.a.log:5757:2018-08-08 06:53:36.902 7f8e12629700  0 mds.0.3 loglog, timeout = 45
$
```

vshankar commented:
> @vshankar What do you think about the behaviour reported in the second paragraph of this comment? I've attached a file in which the console output is copied. I don't think it is acceptable/normal.

I will test out the PR locally. Also, we probably need a test case rather than verifying it manually.

batrick commented:
Any update on this PR?

vshankar commented:
> Any update on this PR?

give me until tomorrow -- will update the pr.

batrick commented:
Needs rebased again

vshankar commented:
@batrick rebased and updated

The hang I mentioned about earlier was not really a deadlock -- the "drop cache" command used to wait for the timeout to expire even if there were no sessions. That's fixed in this update. Also, I removed the custom gather class that didn't do much and used existing `MDSGatherBuilder` class. Plus, added a couple of command related tests.

@batrick I could not reproduce the deadlock issue you had mentioned earlier. Also, w.r.t. the asynchronous execution of "ceph tell" -- I think the command works fine -- although I haven't looked at the tell code closely. I'll try to take a look sometime tomorrow, but for now, the rest of the PR is reviewable.

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -3068,11 +3074,92 @@ bool MDSRankDispatcher::handle_command(
+    }
+
+    JSONFormatter f(true);
+    if (!command_cache_drop(f, timeout)) {
```
I don't see how this will work (`ceph tell`). The dispatcher is processing this `MCommand` message here. I believe it will remain blocked on `command_cache_drop` until the timeout is reached. During this time, the MDS should not be able to process the messages from the clients. (Unless the messenger dispatch thread count is bumped I guess, I've not tried this recently.)

I think it will be necessary to convert this drop cache command into an asynchronous Context which releases the dispatch thread back to the messenger. Consider that this is a problem we share with asynchronous scrub which currently executes in an unaccountable way in the MDS. It'd be valuable to have a generic "command" module of some kind which allows the operator to manipulate/stop these operations once started.

This doesn't need to be written now but it's worth thinking about. I think we can solve `drop cache` by just using contexts to drive the drop to completion and eventually send an MCommandReply once finished. The dispatch thread should be released ASAP. If it simplifies the code, go ahead and drop the asok version.

rishabh-d-dave replied:
batrick [wrote](https://github.com/ceph/ceph/pull/21566#discussion_r215794495)
>  I don't see how this will work (ceph tell). The dispatcher is processing this MCommand message here. I believe it will remain blocked on command_cache_drop until the timeout is reached. During this time, the MDS should not be able to process the messages from the clients. (Unless the messenger dispatch thread count is bumped I guess, I've not tried this recently.)

It indeed does not - 
```
$ date && ./bin/ceph tell mds.a cache drop 20 && date
Fri Sep  7 10:22:51 UTC 2018
{
    "client_recall": {
        "return_code": 0,
        "message": "(0) Success"
    },
    "flush_journal": {
        "return_code": 0,
        "message": ""
    },
    "cache": {
        "pool": {
            "items": 583,
            "bytes": 57080
        }
    }
}
Fri Sep  7 10:22:51 UTC 2018
```

... unlike using the asok command -
```
$ date && ./bin/ceph daemon mds.a cache drop 20 && date
Fri Sep  7 10:19:38 UTC 2018
{
    "client_recall": {
        "return_code": 0,
        "message": "(0) Success"
    },
    "flush_journal": {
        "return_code": 0,
        "message": ""
    },
    "cache": {
        "pool": {
            "items": 729,
            "bytes": 67320
        }
    }
}
Fri Sep  7 10:19:38 UTC 2018
```

The tell command waits until timeout only when `session->caps.size() > newlim`[[1]](https://github.com/ceph/ceph/blob/master/src/mds/Server.cc#L1302) -

```
$ date && ./bin/ceph tell mds.a cache drop 20 && date
Fri Sep  7 10:23:16 UTC 2018
{
    "client_recall": {
        "return_code": 0,
        "message": "(0) Success"
    },
    "flush_journal": {
        "return_code": 0,
        "message": ""
    },
    "cache": {
        "pool": {
            "items": 583,
            "bytes": 57080
        }
    }
}
Fri Sep  7 10:23:17 UTC 2018
```

vshankar replied:
Thanks @batrick -- I understand the issue with `ceph tell` now. With @rishabh-d-dave testing above, I expected `client_recall` status to return `ETIMEDOUT`.

Its probably worth to drive `command_cache_drop()` via context. I'll look into this and update the PR.

vshankar commented:
@batrick Updated PR implements "drop cache" as an asynchronous state machine. This required journal flush to be converted to an asynchronous state machine. The major difference with asynchronous execution is that the `mds_lock` is not held for the entire duration of command execution -- its dropped when invoking an async call and gets reacquired in the context callback. Would that be fine or would we want it to be as it is currently done (in the synchronous version)?

Also, the async implementation might not look elegant as of now -- but I wanted to get this out for reviews and also since we probably want to convert other (tell?) commands to execute asynchronously (for problems similar to what we faced when recalling client state). I have the async state machines as separate commits -- so if we want more discussions/eyes on that, I can resend this with just the asok interface for now.

batrick commented:
@vshankar I still need to look at your update but I want to add onto this PR a note that we should also think about an option to this command to ask clients to drop their object buffers/caches (for all types of clients, kernel and ceph-fuse). This could be a followup ticket or, if you're interested, we can add it to this PR.

vshankar commented:
> @vshankar I still need to look at your update but I want to add onto this PR a note that we should also think about an option to this command to ask clients to drop their object buffers/caches (for all types of clients, kernel and ceph-fuse). This could be a followup ticket or, if you're interested, we can add it to this PR.

I would vote for having that as a follow-up enhancement.

batrick commented:
needs rebase

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -36,11 +36,244 @@
+      gather.activate();
+
+      mdsrank->mds_lock.Unlock();
+      retval = sleeper->wait_for(recall_timeout);
```
We don't want to block the thread at all here. Let's use a `SafeTimer` (? maybe the right choice) which runs its finisher after `recall_timeout`. I think this will require two detachable contexts (like `MDSContextSleeper`). One waits for all clients to trim caps and the other waits for the Timer to complete. We won't be using a condition variable here because we don't want to wait at all. (`MDSContextSleeper` may get axed since I don't think we will use it.)

batrick replied:
Looking at this again, specifically it will block the finisher thread which is a problem. (The dispatcher thread is fortunately free now, which is good!)

vshankar commented on src/mds/MDSRank.cc:
```diff
@@ -36,11 +36,244 @@
+      ss(ss), on_finish(on_finish) {
+  }
+
+  void send() {
```
that's what I got used to when writing async state machines in rbd ;)

vshankar commented on src/mds/MDSRank.cc:
```diff
@@ -36,11 +36,244 @@
 #undef dout_prefix
 #define dout_prefix *_dout << "mds." << whoami << '.' << incarnation << ' '
 
+class C_Flush_Journal : public Context {
```
makes sense

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -36,11 +36,244 @@
 
 #include "MDSRank.h"
 
+#include <boost/scope_exit.hpp>
```
Let's use `src/include/scope_guard.h` instead.

vshankar replied:
this will be removed with the latest update...

vshankar commented:
retest this please

vshankar commented:
@batrick fixed and updated with the following changes

* journal and cache drop async state machines now subclass from `MDSInternalContext`. Also, included are dout logs carried over from earlier implementation.
* waiting for clients (w/ timeout) to trim its cache is now non-blocking -- a custom `C_Drop_Cache::C_ContextTimeout` context is now used to trigger the timeout. this context is also passed to `MDSGatherBuilder` which gets completed if client recall happens before the timeout. also note, that `MDSGatherBuilder` takes care of deleting its finisher (our context callback in this case), so, in the case of client recall timeout, its left to the `gather` class to eventually free this context (later?).
this change gets rid of custom `MDSContextSleeper` class and the changes in `class Cond` for the supporting member functions.
* added tests for dropping cache using `tell` interface.
* added `range=1` for asok and tell interfaces for timeout argument.
